File: medium/322_Coin_Change/coin.go
package main

import "fmt"
import "math"

func max(a, b int) int {
	if a > b {
		return a
	} else {
		return b
	}
}

func coinChange(coins []int, amount int) int {
	f := make([]int, amount+1)
	for i := 1; i < len(f); i++ {
		f[i] = math.MinInt32
	}
	for i := 0; i < len(coins); i++ {
		for j := 0; j <= amount; j++ {
			if j-coins[i] < 0 {
				continue
			}
			tmp := f[j-coins[i]]
			if tmp != math.MinInt32 {
				f[j] = max(f[j], tmp-1)
			}
		}
	}
	if f[amount] == math.MinInt32 {
		return -1
	} else {
		return -f[amount]
	}
}

func testCoin(coins []int, amount int) {
	fmt.Printf("%v, amount %d, get %d\n",
		coins, amount, coinChange(coins, amount))
}

func main() {
	testCoin([]int{1, 2, 5}, 11)
	testCoin([]int{2}, 3)
}

File: medium/322_Coin_Change/coin.py
class Solution(object):
    def coinChange(self, coins, amount):
        """
        :type coins: List[int]
        :type amount: int
        :rtype: int
        """
        dp = [-1 for i in xrange(amount+1)]
        dp[0], minCnt = 0, -1
        for i in xrange(len(coins)):
            for j in xrange(coins[i], amount+1):
                tmp = dp[j-coins[i]]
                if tmp >= 0:
                    if dp[j] == -1:
                        dp[j] = tmp + 1
                    else:
                        dp[j] = min(tmp+1, dp[j])
            if dp[amount] >= 0:
                if minCnt == -1 or dp[amount] < minCnt:
                    minCnt = dp[amount]
        return minCnt
        

class Solution(object):
    def coinChange(self, coins, amount):
        """
        :type coins: List[int]
        :type amount: int
        :rtype: int
        """
        dp = [amount+1 for i in xrange(amount+1)]
        dp[0] = 0
        for i in xrange(len(coins)):
            for j in xrange(coins[i], amount+1):
                dp[j] = min(dp[j-coins[i]]+1, dp[j])
        if dp[amount] > amount:
            return -1
        else:
            return dp[amount]
